### ISSUE TITLE
upload_options + S3 presigning

### ISSUE THREAD

**Issue by reidab** [janko-m/shrine#40]
Looking at [lib/shrine/storage/s3.rb:207](https://github.com/janko-m/shrine/blob/master/lib/shrine/storage/s3.rb#L207), I'd expected options set using the `upload_options` plugin to be taken into consideration when generating a presigned request. However,  [lib/shrine/plugins/upload_options.rb:23](https://github.com/janko-m/shrine/blob/master/lib/shrine/plugins/upload_options.rb#L23) reveals that these options are only applied on `put` to the storage.

I was going to submit a PR making the two work together, but realized the block form of `upload_options` expects an IO object, which we won't have at presigning time… so I'm opening this to document the issue and see you have any thoughts about ways forward.

**Comment by janko-m** [comment]
Yes, unfortunately, in the direct_upload plugin the storage is used directly, and doesn't go through Shrine (it's a storage-specific feature so it doesn't make sense to expose it in `Shrine`), so it's not possible for the `upload_options` plugin to be applied to the presigning in direct_uploads.

However, there are two alternatives. If your upload options aren't dynamic, you can add `:upload_options` when instantiating the storage:

```rb
Shrine::Storage::S3.new(
  upload_options: {acl: "public-read", ...},
  **s3_options
)
```

Alternatively if the upload options need to be dynamic, you can pass a block to `:presign` option in the direct_upload plugin:

```rb
plugin :direct_upload, presign: ->(request) do
  {acl: "public-read"}
end
```

Since you've found `upload_options` options not being upload to `direct_upload` plugin surprising, I'm wondering if those two are connectable but that I'm not seeing a way. Did you have something in mind?

**Comment by reidab** [comment]
Thanks — I ended up using the global upload_options, like you mentioned. The main source of surprise was that [the documentation](http://shrinerb.com/rdoc/classes/Shrine/Storage/S3.html#class-Shrine::Storage::S3-label-Upload+options) seemed to imply that the `upload_options` plugin would work. I think a slight clarification there would probably dodge the issue. :)

I should also admit that I only ran across this by following a somewhat non-standard workflow. I'm implementing presigning in an API that uses [JSON-API](http://jsonapi.org) doesn't serve any HTML views, so I needed to reimplement the functionality provided by the endpoint in the `direct_upload` plugin.
